### PR TITLE
Link to local api docs in installer included docs

### LINKED ISSUE #728: Point External Link at Different Location in Compile Time
Include content based on configuration
https://www.sphinx-doc.org/en/master/usage/extensions/ifconfig.html

### PR DESCRIPTION
**Pull request by TheTripleV**
closes #728 .

If docs are accessed from the wpilib install location, the links to the Java/C++ Api Docs will go to local versions instead of online versions.

Tested on Mac.

### CODE REVIEW

File: source/cpp-api-docs-redirect.rst [hunk 1]
```diff
@@ -0,0 +1,18 @@
+C++ Api Documentation
+======================
```
Comment (Daltz333) [inline source/cpp-api-docs-redirect.rst:2]:
```suggestion
=====================
```

**Comment by TheTripleV** [comment]
rebased off master

**Comment by Daltz333** [comment]
Not sure if it's possible to add a check to confirm that the destination exists (locally), for cases where the RTD docs is manually downloaded

**Comment by Daltz333** [comment]
I have confirmed that this works on Windows

**Review by prateekma — CHANGES_REQUESTED** [review]
Change Api -> API

File: source/cpp-api-docs-redirect.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+            window.location.href = documentation_url_base + sep + "cpp" + sep + "index.html";
+        } else {
+            window.location.href = "https://first.wpi.edu/FRC/roborio/release/docs/cpp/index.html";
+        }
+    </script>
+
+.. image:: assets/cat-redirect.jpg
+   :alt: Cat Yawning
+
+Oops. Your browser doesn't support Javascript.
```
Comment (prateekma) [inline source/cpp-api-docs-redirect.rst:19]: Javascript -> JavaScript

**Review by prateekma — CHANGES_REQUESTED** [review]

**Review by Daltz333 — DISMISSED** [review]

**Comment by TheTripleV** [comment]
The links to the api docs in `docs/software/advanced-controls/trajectories/ramsete.rst` also need to be fixed.
They link to specific pages in the api docs while this PR currently only supports linking to the root of the api docs.

Solution: 
In ramsete.rst, we can link to the built redirect html file instead of the rst. The desired path (relative to the root of the api docs) can be passed in the query portion of the url and the js can redirect correctly.

**Comment by sciencewhiz** [comment]
I believe there are many more links to the api docs, especially in the command section.

**Comment by TheTripleV** [comment]
I think all links can be automatically taken care of using sphinx's build-finished event.

1. The redirecting file will be updated to allow for redirecting to any page within the api docs.
2. On build-finished, all links to the api docs in frc-docs will be replaced with redirecting ones.

The only downside I see to this solution as opposed to redirecting in the wpilib installer is that this solution requires that javascript is enabled.

So I guess the decision is where to do the find and replace, here or in the wpilib installer?

**Comment by Daltz333** [comment]
It can be done via JavaScript. I don't think we need to support a no-javascript environment, as the website is already severely broken without it.

File: source/java-api-docs-redirect.rst [hunk 3]
```diff
@@ -1,16 +1,13 @@
-Java API Documentation
-======================
+WPILIB Java API Documentation
```
Comment (Daltz333) [inline source/java-api-docs-redirect.rst:1]: WPILib

**Comment by TheTripleV** [comment]
Changes:
- All links to the api docs from frc-docs are properly redirected, including links to specific api pages
- The hash is propagated to the redirected url
- Code duplication reduced - the main redirecting js code only exists once now.

Functionality:
All links to the api docs are replaced with a link to a dummy file that conditionally redirects to local/online versions of the api docs. Links to the api docs of form `https://first.wpi.edu/FRC/roborio/release/docs/LANG/PAGE#HASH` are replaced with `api-docs-redirect.html?path=LANG/PAGE&hash=HASH` which is then parsed by the js in `api-docs-redirect.rst`

Note: When `index.rst` is read, `api-docs-redirect.html` has not been built yet, resulting in an error. This is why `cpp-api-docs-redirect.rst` and `java-api-docs-redirect.rst` exist. Users clicking on the api docs links in the nav sidebar will face a double redirect while all other api docs links only result in a single redirect.

**Comment by TheTripleV** [comment]
Because of the double redirect on the sidebar, the browser back doesn't work. Need to implement another workaround. Will do after AoC.

**Comment by TheTripleV** [comment]
Changes:
The dummy redirecting page has been removed. Now, on page load, all api docs links are changed in the DOM. This makes the redirects "instant" and doesn't mangle the browser history.

File: source/_static/js/api-docs-redirect.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+document.addEventListener('DOMContentLoaded', function() {
+    "use strict";
+    for (let link of document.links) {
+        if (link.href.startsWith("https://first.wpi.edu/FRC/roborio/release/docs/")) {
+            link.href = resolveApiDocsLink(link.href)
+        }
+    }
+}, false);
```
Comment (AustinShalit) [inline source/_static/js/api-docs-redirect.js:24]:
```suggestion
}, false);

```

**Review by Daltz333 — APPROVED** [review]

**Pull request merged**